### PR TITLE
Disable Rebuild Data Source Button when one is already in progress

### PR DESCRIPTION
**Pull request by AmitPhulera**
## Product Description

https://dimagi-dev.atlassian.net/browse/SS-220
We are disabling the Rebuild Data Source button when one is already in progress. Cal has earlier worked on doing it at the task level in  https://github.com/dimagi/commcare-hq/pull/30288 now we are adding the feedback to the UI.
After these changes when rebuild would be in progress the button will look like 

Screenshot 2021-10-05 at 6 55 19 PM

Screenshot 2021-10-05 at 6 55 08 PM

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### QA Plan

Will add later.

### Safety story

These are majorly cosmetic changes and have been tested on local. I will be putting it on staging after getting initial review. 
### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/userreports/views.py [hunk 1]
```diff
@@ -1251,13 +1261,19 @@ def build_data_source_in_place(request, domain, config_id):
...
+    _set_ucr_cache_key(config_id)
     rebuild_indicators_in_place.delay(config_id, request.user.username, source='edit_data_source_build_in_place')
     return HttpResponseRedirect(reverse(
         EditDataSourceView.urlname, args=[domain, config._id]
     ))
 
 
+def _set_ucr_cache_key(config_id):
+    key = DATA_SOURCE_REBUILD_CACHE_PREFIX + config_id
+    CACHE_TTL = 24 * 60 * 60
```
Comment (AmitPhulera) [inline corehq/apps/userreports/views.py:1273]: Hey @calellowitz !
What do you think here ideal TTL for the cache should be?

**Review by biyeun — APPROVED** [review]
looking good. nice!

**Comment by AmitPhulera** [comment]
Hey @calellowitz, 

Thanks for your review. I have a few things to ask/share on the things that you have mentioned.

> We use it to display a banner to the user in the report ui in https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/templates/userreports/partials/build_in_progress_warning.html

Thanks for sharing this, I will update the code to use this instead.

> There is one other thing to consider that makes me less sure we even want to make this change. If a user makes a change, kicks of a rebuild, then proceeds to make more changes, I am not certain we want to prevent them from pressing the rebuild button.

Is the general workflow on how the datasources are configured? If yes then I totally agree that holding off user would be a bad idea.

> In the case where this kind of thing matters, rebuilds that take a long time, it is a pretty terrible user experience if to reflect your changes, you need to wait an unknown amount of time (periodically checking in during that time to see if its ready) and then hit the rebuild button, perhaps hours after you finished making changes. You want to be able to kick it off (even if it doesn't run immediately), when you are done working, not at some vague point in the future. I don't think many people are repeatedly hitting the button without changes, they are hitting it as they make changes, then making new changes and hitting it again.

Let's assume that we don't change the workflow then do you think that we should convey this to user that their rebuild will start automatically when the previous one is finished?

> Although some of that only works with report builder. It might be good to try to make that functionality work with rebuilds triggered from this UI as well.

I did not quite understand it. Can you please elaborate?

**Review by calellowitz — CHANGES_REQUESTED** [review]
What does this look like? I was imagining a help text like your last message, but it looks like this is slightly different. Do you have screen shots?

Also,  can you make https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/tasks.py#L128-L129, so that it's behavior matches the other type of rebuild (that was my oversight last time I worked on this).

**Comment by AmitPhulera** [comment]
>  What does this look like? I was imagining a help text like your last message, but it looks like this is slightly different. Do you have screen shots?

As soon as you will hover over the button, you will see something like this.

Screenshot 2021-10-14 at 11 03 17 AM


When a rebuild is happening then the help box will appear. 

Screenshot 2021-10-14 at 11 03 46 AM

The reason, why I am showing a warning on hover, is so that users get to know about what rebuilds are before they click it. If it would have been in the help button then they would have to click/hover over help to understand what it actually says.


> Also, can you make https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/tasks.py#L128-L129, so that it's behavior matches the other type of rebuild (that was my oversight last time I worked on this).

Sure. I will do it.

**Review by calellowitz — APPROVED** [review]

**Review by calellowitz — APPROVED** [review]

**Pull request merged**